scripts/wrapper/withdraw.js

```diff
@@ -48,25 +77,50 @@ module.exports = function (web3, artifacts) {
 
     assertGoodArguments(argv)
 
-    let withdrawals = JSON.parse(await fs.readFile(argv.withdrawalFile, "utf8"))
-    const tokenInfoPromises = fetchTokenInfoForFlux(withdrawals)
-    const exchange = await getExchange(web3)
+    let withdrawals
+    let tokenInfoPromises
+    if (argv.withdrawalFile) {
+      withdrawals = JSON.parse(await fs.readFile(argv.withdrawalFile, "utf8"))
+      tokenInfoPromises = fetchTokenInfoForFlux(withdrawals)
+    } else {
+      const exchangePromise = getExchange(web3)
+      if (argv.tokens) tokenInfoPromises = fetchTokenInfoAtAddresses(argv.tokens)
+      else {
+        const tokenAddresses = await Promise.all(
+          argv.tokenIds.map(async (id) => (await exchangePromise).tokenIdToAddressMap(id))
+        )
+        tokenInfoPromises = fetchTokenInfoAtAddresses(tokenAddresses)
+      }
+      const exchange = await exchangePromise
+      const currentBatchId = (await exchange.getCurrentBatchId()).toNumber() // cannot be computed directly from Date() because time of testing blockchain is not consistent with system clock
 
-    if (argv.allTokens) {
       log("Retrieving amount of tokens to withdraw.")
       // get full amount to withdraw from the blockchain
-      withdrawals = await Promise.all(
-        withdrawals.map(async (withdrawal) => ({
-          bracketAddress: withdrawal.bracketAddress,
-          tokenAddress: withdrawal.tokenAddress,
-          amount: await getMaxWithdrawableAmount(
-            argv,
-            withdrawal.bracketAddress,
-            await tokenInfoPromises[withdrawal.tokenAddress],
-            await exchange
-          ),
-        }))
-      )
+      withdrawals = []
+      const candidateWithdrawalPromises = []
+      for (const [, tokenDataPromise] of Object.entries(tokenInfoPromises))
```

Comment:
I think the script would be easier to understand if you would await all tokenInfoPromises before going over them in this loop. Then you don't get promises of promises, and the performance lost would be negligible, right?

If I am not mistaken, then you would not even need
```
 for (const candidateWithdrawalPromise of candidateWithdrawalPromises)
```
then? But I am not sure right now.

Reply:
I tried to make the code more linear with the new changes. I still couldn't avoid creating some intermediate promises to populate the withdrawal list, but there are less nested calls so it should be more readable.